Refactor hunger and thirst systems

Centronias opened the pull request:
## About the PR

Spiritual successor to #27481 .
Thank you to @Lukasz825700516 for figuring out the initial stuff, I definitely wouldn't've been able to conceptualize this all on my own.

Combines hunger and thirst into one concept, "satiation", which is extensible to new satiation types via yaml-only
changes.
Also gets rid of the C#-defined threshold enums.
Additionally, this applies the same network-burden-reducing changes from #32986 (with what're hopefully even some
improvements, see below).

Supercedes #33403 .

Also note that this steps on changes to make Thirst and Hunger use field deltas in #28242 and #??? respectively.

## Why / Balance

- Lots of shared code.
- Got like three people saying "why not just do #27481 ?" on my #32986 review.
- Killing hardcoded enums good; more power to the yamlers

## Technical details

### Prediction of linear decay of satiations

This PR includes a lot of the changes from #32986 ,which are intended to reduce how many component state changes are
sent over the network. This basically means that instead of the component having a `currentValue` that's updated and
synchronized on every tick, getting the "current value" is instead a calculation which determines the predicted value
based on an initial value, when that initial value was set, a constant decay rate, and the current time.

In addition to the changes related to predicting decay, this PR includes an additional optimization. The implementation
in #32986 checks if the predicted current value has crossed a threshold. This implementation instead calculates **when
that threshold will be crossed when setting the value, and then waits until that time to check**. This difference saves
needing to resolve the satiation prototype, look up the threshold, etc. on each update call.

### Caching

A big change in this is that thresholds are entirely flexibly defined in YAML as opposed to being C# enums. Broadly the
way that this works is that the assorted values in a `SatiationPrototype` are all keyed by a satiation value at which they
become active, and every unique key-value is treated as a threshold. For example, if want a satiation with behavior which
looks like this

| Satiation Value<br/>(Effects active at and below this<br/>value until next threshold reached) | Speed Modifier  | Decay Modifier  |
|-----------------------------------------------------------------------------------------------|-----------------|-----------------|
| `100`                                                                                         | `1.0` (Default) | `1.2`           |
| `67`                                                                                          | `8.0`           | `1.0` (Default) |
| `33`                                                                                          | `8.0`           | `0.8`           |

The `SatiationPrototype`'s YAML would look like this:

```yaml
- type: satiation
  # Superfluous fields omitted.
  speedModifiers:
    # Everything above the highest key uses the default modifier of 1.0
    67: 8.0
    # Everything below the lowest key uses that key's value
  decayModifiers:
    100: 1.2
    67: 1.0 # Need to explicitly "stop" the `1.2` value from above
    33: 0.8
```

(Note that all of the key-values here can be key strings instead: see [below](#get-rid-of-threshold-enums-in-favor-of-yaml-defined-keys)).

When `SatiationSystem` initializes and whenever `SatiationPrototype`s are modified, caching is performed to collate all of the threshold
satiation values and their associated effect values into one unified ordered list of "When you're at this value, these are all of the
effects applied to you". It ends up looking very similar to the first table above as there is one element per threshold value where things
change, and all of the effect values are filled in, even if they are their default values.

Check out the implementation in `SatiationSystem.Caching.cs`; I've tried to document all of the parts I used and the sticky parts of the
actual implementation, but I am sure it is still not the clearest thing in the world. One of the tricky parts was that it is important to
differentiate between a default/null value for an effect versus an absent/undefined value at a threshold, so that's why I added a `Maybe`
type in that file 🤡.

### Code Changes

#### Core Changes:

- Remove
    - `HungerComponent` and `ThirstComponent`
    - `HungerThreshold` and `ThirstThreshold`
    - `HungerSystem` and `ThirstSystem`
    - `HungerThreshold` and `ThirstThreshold`
- Add:
    - `SatiationComponent`
        - Basically just a map of `SatiationTypePrototype` to `Satiation`
    - `Satiation`
        - The thing that's actually comparable to the old `HungerComponent` and `ThirstComponent`
        - Holds onto the state of a single satiation in the form of a `LastAuthoritativeValue` and
          `LastAuthoritativeChangeTime`, as in #32986
        - Also holds onto a handful of values which are used in regular updates so as to avoid needing to recalculate
          those values on each update tick.
    - `SatiationTypePrototype`
        - A definition of a _type_ of satiation
        - Currently just `Hunger` and `Thirst` are defined
    - `SatiationPrototype`
        - A definition of a satiation
        - Instead of defining thresholds and values on mob prototypes, you define a `SatiationPrototype` and reference
          it in the mob's proto
        - Examples include `DionaHunger`, `SnailHunger`, `RegalRatThirst`
        - The definition of [keys](#get-rid-of-threshold-enums-in-favor-of-yaml-defined-keys) valid for this satiation
    - `SatiationValue`
        - A value used to define at what satiation values certain effects on `SatiationPrototypes` become active
        - This is the replacement for the old `HungerThreshold` and `ThirstThreshold` enums.
          See [below](#get-rid-of-threshold-enums-in-favor-of-yaml-defined-keys) for more.
        - Is either a direct value or a key string
            - When using a key string, the key is looked up in `SatiationProtoype.Keys` to determine the integer value
              that key represents
        - A custom type reader is implemented for `SatiationValue` so that yamlers need not utilize `!type:` tags.
          Simply use an integer when you want to specify the value directly, or a string when you want to use a key.
    - `SatiationSystem`
        - Combined replacement of the old systems
        - Handles events:
            - `MapInitEvent`: initialize satiations with `Okay` values
            - `ComponentShutdown`: clears alerts when satiation is removed
            - `EntityUnpausedEvent`: deals with `TimeSpan` fields in satiations
            - `RefreshMovementSpeedModifiersEvent` applied movespeed modifications based on satiation levels
            - `RejuvenateEvent`: sets satiations to maximum values
            - `PrototypesReloadedEventArgs`: reacts to changes to satiation prototypes for caching purposes
        - Implements `Update`:
            - Iterates through each `Satiation` in the component, doing the following:
            - Continuously apply continuous effects (just damage, when defined on the prototype)
            - Predictively handles transitions from one threshold to another, updating value, effects, and decay rate
        - Provides an "API" of public methods to interact with satiations from other systems and commands (see
          `SatiationSystem.Api.cs` and `SatiationSystem.Commands.cs` in the system):
            - `float? GetValueOrNull(Entity<SatiationComponent>, ProtoId<SatiationTypePrototype>)`
            - `void SetValue(Entity<SatiationComponent>, ProtoId<SatiationTypePrototype>, SatiationValue)`
            - `void SetValue(Entity<SatiationComponent>, ProtoId<SatiationTypePrototype>, float value)`
            - `void ModifyValue(Entity<SatiationComponent>, ProtoId<SatiationTypePrototype>, float amount)`
            - `bool IsValueInRange(Entity<SatiationComponent>, ProtoId<SatiationTypePrototype>, SatiationValue? above = null,
        SatiationValue? below = null, float hypotheticalValueDelta`
        - Provides some additional accessors for other more specific needs:
        - `bool TryGetValueByThreshold<T>(Entity<SatiationComponent>, ProtoId<SatiationTypePrototype>, Dictionary<string, T>, out T? result)`
        - and `bool TryGetValueByThreshold<T>(Entity<SatiationComponent>, ProtoId<SatiationTypePrototype>, Dictionary<int, T>, out T? result)`
          - Not used publicly, but exposes a powerful capability which the system uses to retrieve values from a table based on the current
            satiation value. This is how effects are retrieved from the [cache](#caching) internally.
        - `StatusIconPrototype? GetStatusIconOrNull(Entity<SatiationComponent>, ProtoId<SatiationTypePrototype>)`
          - Used by `ShowSatiationIconsSystem` only
        - `SatiationTypePrototype? GetTypeOrNull(string protoId)`
          - Used by commands to get / validate user-entered satiation types
        - `IEnumerable<string> GetKeysForType(Entity<SatiationComponent>, ProtoId<SatiationTypePrototype>)`
          - Used by commands to get valid key strings for a specific entity & satiation
        - `int? GetMaximumValue(Entity<SatiationComponent> entity, ProtoId<SatiationTypePrototype>)`
          - Used by commands to suggest a numeric satiation value for a specific entity & satiation
        - System behavior boils down to two methods:
          - `float CalculateCurrentValue(Satiation, SatiationPrototype)`
            - Gets the value of a satiation based on its authoritative time and value, and the current time.
            - This is where the prediction of linear decay happens.
          - `void SetAuthoritativeValue(Entity<SatiationComponent>, Satiation, SatiationPrototype, float value)`
            - Sets the value of a satiation, including all of the effects that has (setting authoritative values,
              updating threshold, updating threshold effects, etc.)

#### Get rid of Threshold Enums in favor of YAML-defined Keys

Previously, Hunger and Thirst used C# enums to define thresholds and associate effects with those thresholds. For example, when `Peckish`,
an entity would move more slowly. This required the hunger prototype to define the numeric values which corresponded to being in the
`Peckish` state. This was useful for flexibly defining hunger across different entities where a mouse would be `Peckish` at 25 hunger while
a human would be 100 hunger (something something body size relative to nutrition in food). The neat thing was that the "you get slowed down
while `Peckish`" only needed to be defined once, and it was just the numbers associated with `Peckish` that needed to change.

I wanted to maintain that flexibility, so I have implemented `Keys` on prototypes. The work basically the same as the threshold enums except
that they're defined entirely in YAML and they're a bit more squishy. You can define as many keys as you want on a `SatiationPrototype`, and
they're really just a shorthand for a numeric value. The flexibility comes in with prototype inheritance. A base `SatiationPrototype` can
define the effects (e.g. the slowdown like before) of being at an abstract "`Peckish`" satiation level, and then the inheriting
`SatiationPrototype`s assign numeric values to that abstract `Peckish` key. Here's a concrete YAML example:

```yaml
- type: satiation
  id: BaseSatiation
  # Superfluous fields omitted.
  keys:
    Full: 200
    Okay: 150
    Low: 100
    Dead: 0
  startingValueMinimum: 0    # Most fields can use either a literal numeric value or refer to a value by its
  startingValueMaximum: Okay # Key string. Both of these `startingValue` fields are of the same C# type, but
  # can take either an int or a string in yaml!
  icons:
    Full: DemoIconFull # When between Okay and Full, `DemoIconFull` is used.
    Okay: null         # Explicitly assigning `null` here means when between Low and Okay, no icon is shown
    Low: DemoIconLow
    Dead: DemoIconDead

- type: satiation
  # Superfluous fields omitted.
  parent: BaseSatiation
  id: IconOnlyWhenDeadSatiation
  # Note we don't need to redefine the keys if we only want to change values!
  icons: # Overwrite _all_ of the icon definitions with this one.
    Dead: DemoIcon # Only show an icon when at or below Dead satiation.

- type: satiation
  # Superfluous fields omitted.
  parent: BaseSatiation
  ids: AlwaysLowSatiation
  keys: # Redefine the keys so that the Low state is really big!
    Full: 200
    Okay: 195
    Low: 190
    Dead: 0
  # We don't need to redefine the values because we inherit them! All we did was change when they're
  # used by modifying the keys!
```

Because the enums were used in a few places across the codebase, I had to modify a handful of components to use keys define in YAML instead:

- `DrinkValueCon`
- `FoodValueCon`
- `FatExtractorComponent`
- `UdderComponent`
- `WoolyComponent`
- `ReproductiveComponent`
- `SericultureComponent`

#### Add new Tests

- `SatiationTest`: more or less just verifies the API functions do what they say.
- `SatiationCachingTest`: uses a backdoor into `SatiationSystem`'s threshold cache to make sure thresholds work as
  described [above](#caching)

#### Consolidate other stuff particular to Hunger or Thirst to shared Satiation implementations:

- `ShowHungerIconsSystem` and `ShowThirstIconSystem` -> `ShowSatiationIconsSystem`
  - Add `ShowSatiationIconsComponent`
- `HungerCondition` EntityCondition -> `SatiationCondition` EntityCondition
- `SatiateHunger`, `SatiateThirst` and abstract `Satiate<T>` EntityEffects -> `Satiate` EntityEffect
- `HungryPrecondition` and `ThirstyPrecondition` HTNPreconditions -> `SatiationPrecondition` HTNPrecondition
- Unused `Hunger` and `Thirst` log types -> equally unused `Satiation` log type
- `ExaminableHungerComponent` -> `ExaminableSatiationComponent`
- Various usages in systems

#### Console Command Changes:

- Get rid of `hungry` and `thirsty`, replaced by `unsatiate`
- Replace `setnutrit` with `setsatiation`
- (Also replace a bunch of unique localization strings for these commands with common ones)

#### Lots of Prototype Changes:

- `Hunger` and `Thirst` satiation type prototypes
- Hunger Prototypes:
  - `NormalSatiationHunger` (Holds old implementation's hard-coded defaults)
  - `SimpleMobBaseHunger` (Used by most animals)
  - `SnailHunger`
  - `CockroachHunger`
  - `MouseHunger`
  - `MouseAdmemeHunger`
  - `HamsterHunger`
  - `RegalRatHunger` (Ratking + Servants)
  - `DionaHunger` (trees don't need to eat as much)
- Thirst Prototypes:
  - `NormalSatiationThirst` (Hold old implementation's hard-coded defaults)
  - `MouseThirst`
  - `MouseAdmemeThirst`
  - `SimpleMobBaseThirst` (Used by most animals)
  - `DionaThirst` (again, trees OP)
- Change mob hunger and thirst to use new prototypes
- Replace a lot of Hunger/Thirst `EntityEffectCondition`s, `EntityEffect`s, and `HTNPrecondition`s with the
  aforementioned combined ones.

## Media

**Normal Satiation stuff, and showing off the authoritative values and predicted threshold change time**

https://github.com/user-attachments/assets/46970eb6-f8c0-438a-982d-761148c24ff7

**Console commands with completion**

https://github.com/user-attachments/assets/de4850f2-b599-41c1-8a31-e48ecb72c5ef

**NPC Mouse eats**

https://github.com/user-attachments/assets/6fa52299-9f10-405c-83ba-4c67c2c87869

**HUD Icons**

https://github.com/user-attachments/assets/b05a9271-813f-4176-a3ff-92f9c6290792

** Examinable Satiation**
![image](https://github.com/user-attachments/assets/13dcf92e-f460-4e02-af2b-9a16f56550e0)

**`setsatiation` Completion**
<img width="597" height="273" alt="image" src="https://github.com/user-attachments/assets/a0b91ca2-508e-4940-bce2-130cb7e5f3ea" />

## Requirements

- [X] I have read and am following
  the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes

#### Changes to code

`HungerComponent`, `HungerThreshold`, `HungerSystem`, `ThirstComponent`, `ThirstThreshold`, and `ThirstSystem` have been removed, and
getting & setting hunger and thirst now go through the `SatiationComponent` and the `SatiationSystem`. Note that in a lot of these examples,
I just use `"Hunger"` and `"Thirst"` which are the IDs of `SatiationTypePrototype`s. In your usage, you use use `SatiationSystem.Hunger` and
`SatiationSystem.Thirst`, which are validated `ProtoId<SatiationType>`s for these values.

##### Refer to Hunger and Thirst Thresholds

Previously, `HungerThreshold` and `ThirstThreshold` were enums defined in C# and which were "states" in which hunger and thirst,
respectively, could be in depending on the actual value of hunger and thirst. It was possible to interrogate the components to see what
threshold their respective values were at and modify system behavior accordingly (see [Check Threshold below](#check-threshold)).

Now, there is no concept of satiations being "in" a threshold, and instead we only ever deal with the satiation value being above, below, or
between certain values. What this means is that if your system was directly referencing a threshold enum value before, it will need to use
a value or a Satiation Value Key instead (see [Defining Entities with Satiation Needs below](#defining-entities-with-satiation-needs)).

##### Get Current Values

```csharp
// Before
var hunger = Comp<HungerComponent>(uid);
float currentHunger = _hunger.GetHunger(hunger);

var thirst = Comp<ThirstComponent>(uid);
float currentThirst = thirst.currentValue;

// After
SatiationSystem satiationSys = ...;
var satiation = Comp<SatiationComponent>(uid);
float? currentHunger = satiationSys.GetValueOrNull((uid, satiation), "Hunger");
float? currentThirst = satiationSys.GetValueOrNull((uid, satiation), "Thirst");
// Note that the values are nullable because the given satiation type may not be present in the given satiation component.
```

##### Modify Values

```csharp
float change = -1.0;

// Before
HungerSystem hungerSys = ...;
var hunger = Comp<HungerComponent>(uid);
hungerSys.ModifyHunger(uid, change, hunger);

ThirstSystem thirstSys = ...;
var thirst = Comp<ThirstComponent>(uid);
thirstSystem.ModifyThirst(uid, thirst, change);

// After
SatiationSystem satiationSys = ...;
var satiation = Comp<SatiationComponent>(uid);
satiationSys.ModifyValue((uid, satiation), "Hunger", change);
satiationSys.ModifyValue((uid, satiation), "Thirst", change);
```

##### Check Threshold

(Used by Sericulture, Wooly, and Udder systems -- check if a modification to a satiation would put it into a different threshold)

```csharp
float change = -1.0;

// Before
HungerThreshold neededThreshold = ...;
HungerSystem hungerSys = ...;
var hunger = Comp<HungerComponent>(uid);
bool wouldBeBelowThresholdAfterChange = hungerSys.IsHungerBelowState(uid, neededThreshold, hungerSys.GetHunger(hunger) + change, hungerComp);

// After
string neededThreshold = ...;
SatiationSystem satiationSys = ...;
var satiation = Comp<SatiationComponent>(uid);
bool wouldBeBelowThresholdAfterChange = satiationSys.IsValueInRange((uid, satiation), "Hunger", above: neededThreshold, hypotheticalValueDelta: change);
```

#### Changes to yaml

##### Using Satiation Value Keys in YAML

Previously, `HungerThreshold` and `ThirstThreshold` were enums defined in C# and which were "states" in which hunger and thirst,
respectively, could be in depending on the actual value of hunger and thirst. YAML could refer to these thresholds by their names (e.g.
`OverHydrated`, `Overfed`, `Okay`, etc.) and the underlying C# would know what to do with them.

Now, there is no concept of satiations thresholds, and instead of referring to threshold enum values, we instead define and refer to
**Satiation Value Keys**. Keys are defined on a `SatiationPrototype` in the `keys` field (shocking!) and they are simply a mapping of key
string to value (see [Defining Entities with Satiation Needs below](#defining-entities-with-satiation-needs) for an example). Satiation
Value Keys provide a lot of the same definitional power as thresholds did before: many fields on `SatiationPrototype`s will look the same as
they did before because keys can be used to refer to breakpoints of satiations by name. The additional power that Satiation Value Keys
provide, though, are that you can add as many as you want, and even add specific "marker" keys for particular systems and functions.

As long as the `key` field is not overwritten by an inheritor, prototypes will inherit keys from their parent prototypes, allowing for reuse
and concise definitions of `SatiationPrototypes`. The corollary of this is that fields are inherited as normal, so by defining
`speedModifiers`, `decayModifiers`, etc. using keys, it is possible for an inheritor to modify only the Satiation Value Key values to
reuse the values while changing when they take effect.

The following pitfalls should be kept in mind and avoided, however:
- Keys are case sensitive
- A key which is referred to but not defined is silently ignored
- Keys are defined _per prototype_, so trying to use `OverHydrated` on `SimpleMobBaseHunger` will be ignored

##### Defining Entities with Satiation Needs

Because the satiations have been combined, defining entities with hunger and thirst has changed substantially. Before,
the hunger and thirst components would be defined directly and separately on an entity. Now, they're combined into
`satiation`, and rather than defining most parts directly, they instead refer to a prototype.

Additionally, various fields were implicitly defined based on their threshold values:
- the maximum value of a satiation was previously implicitly defined by the `Overfed` and `OverHydrated` threshold values, whereas now it is 
  necessary to explicitly define the maximum value of a satiation with `maximumValue`. (This MAY NOT be a Satiation Value Key)
- the starting values of satiations were implicitly in the `Okay` threshold whereas now it is necessary to define the starting range using
  the `startingValueMinimum` and `startingValueMaximum` values. (These MAY be a Satiation Value Key)
- the slowdown from low hunger or thirst was implicitly applied when in and below the `Parched` and `Starving` thresholds. Now you can apply
  arbitrary speed modifiers using `speedModifiers`.


```yaml
# Before ====================
- type: entity
  name: mouse
  parent: SimpleMobBase
  id: MobMouse
  components:
  # Lots of other components...
  - type: Thirst
    startingThirst: 25
    thresholds:
      OverHydrated: 35
      Okay: 25
      Thirsty: 15
      Parched: 10
      Dead: 0
    baseDecayRate: 0.04
  - type: Hunger
    thresholds:
      Overfed: 35
      Okay: 25
      Peckish: 15
      Starving: 10
      Dead: 0
    baseDecayRate: 0.5

# After ====================
# Define the entity
- type: entity
  name: mouse
  parent: SimpleMobBase
  id: MobMouse
  components:
  - type: Satiation
    satiations:
      Hunger: # <--- Key here refers to a `satiationType`
        prototype: MouseHunger # <--- This proto ID refers to a `satiation` 
      Thirst:
        prototype: MouseThirst

# Define satiation types
- type: satiationType
  id: Hunger
  name: satiation-type-hunger

- type: satiationType
  id: Thirst
  name: satiation-type-thirst

# Define satiations
- type: satiation
  id: MouseHunger
  # parent: ...
  #         Note that most of this would be inherited from a parent, in normal usage.
  #         I've just inlined it all here for full illustration  
  baseDecayRate: 0.5
  maximum: 35
  startingValueMinimum: 20
  startingValueMaximum: Okay
  keys: # Keys MUST be defined here otherwise their values in other sections will be ignored.
    Overfed: 35
    Okay: 25
    Peckish: 15
    Starving: 10
    Dead: 0
  decayModifiers:
    Overfed: 1.2
    Okay: 1.0
    Peckish: 0.8
    Starving: 0.6
    Dead: 0.6
  speedModifiers:
    Starving: 0.75
  alertCategory: Hunger
  alerts:
    Peckish: Peckish
    Starving: Starving
    Dead: Starving
  icons:
    Overfed: HungerIconOverfed
    Peckish: HungerIconPeckish
    Starving: HungerIconStarving
    Dead: HungerIconStarving
```

##### `ExaminableHunger ` component changed to `ExaminableSatiation`

Before the ability to examine hunger was hard-coded to hunger. Now it works with arbitrary satiations, but you need to define some
prototypes to hook it up.

```yaml
# Before ====================
- type: entity
  name: chicken
  parent: SimpleMobBase
  id: MobChicken
  components:
  - type: ExaminableHunger
    # Note that on Wizden, there were implicitly specified localization descriptions. under this component

# After ====================

# Define a SatiationExamination prototype (Localization string definitions not shown in the example)
- type: satiationExamination
  id: HungerExamination
  descriptions: # <--- Keys in this object are Satiation Value Keys
    Overfed: "examinable-satiation-component-examine-hunger-full"
    Okay: "examinable-satiation-component-examine-hunger-okay"
    Peckish: "examinable-satiation-component-examine-hunger-concerned"
    Starving: "examinable-satiation-component-examine-hunger-desperate"
    Dead: "examinable-satiation-component-examine-hunger-desperate"
  notApplicable: "examinable-satiation-component-examine-hunger-none"

# Put the component on the entity, referencing the SatiationExamination
- type: entity
  name: chicken
  parent: SimpleMobBase
  id: MobChicken
  components:
  - type: ExaminableSatiation
    satiations:
      Hunger: HungerExamination
```

##### Entity Condition `HungerCondition` changed to `SatiationCondition`

Before, the application of entity effects could be predicated on the hunger of an entity. This is still possible, but
now it also enables checking other satiation types, and requires the condition definition include the satiation type
checked.

```yaml
# Before
effects:
- !type:HealthChange
  conditions:
  - !type:HungerCondition
    max: 50
  damage:
    types:
      Poison: 2

# After
effects:
- !type:HealthChange
  conditions:
  - !type:SatiationCondition
    satiationType: Hunger # <--- Refers to a `satiationType` as defined elsewhere
    max: 50
  damage:
    types:
      Poison: 2
```

(Basically find anywhere you use `!type:HungerCondition` and replace it with `!type:SatiationCondition \n satiationType: Hunger`. Just
make sure the whitespace is correct :) )

##### Entity Effects `SatiateHunger` and `SatiateThirst` changed to `Satiate`

Hunger and thirst can still be modified by entity effects, but rather than having separate distinct effect definitions,
they now share one and that one needs to have the affected satiation defined.

```yaml
# Before
effects:
- !type:SatiateThirst
  factor: -1.5

# After
effects:
- !type:Satiate
  satiationType: Thirst # <--- Refers to a `satiationType` as defined elsewhere
  factor: -1.5
```

(Basically find anywhere you use `!type:SatiateHunger` and replace it with `!type:Satiation \n satiationType: Hunger`,
and same with thirst. Just make sure the whitespace is correct :) )

##### HTN Preconditions `HungryPrecondition` and `ThirstyPrecondition` changed to `SatiationPrecondition`

Hunger and thirst can still be used as preconditions for HTN tasks, but rather than having separate distinct
precondition definitions, they now share one that needs to have the checked satiation specified. Additionally,
previously the `minHungerState` and `minThirstState` values implicitly referred to the "bottom" of that threshold
(because they were minimums), and now we need to refer to a specific range of values instead using the `above` and
`below` fields.

```yaml
# Before
- !type:HTNPrimitiveTask
  preconditions:
  - !type:HungryPrecondition
    minHungerState: Starving # <--- Refers to `HungerThreshold`, which was removed

# After
- !type:HTNPrimitiveTask
  preconditions:
  - !type:SatiationPrecondition
    satiationType: Hunger # <--- Refers to a `satiationType` as defined elsewhere
    below: Starving # <--- Refers to a Satiation Value Key
```

##### NPC Utility `FoodValueCon` and `DrinkValueCon` are Configurable

Previously, they had hardcoded hunger and thirst thresholds at which their considerations returned `0`. Now you can
configure how hungry or thirsty the NPC has to be for the consideration to return `0` or its curve value.

```yaml
# Before
- !type:FoodValueCon
  curve: !type:QuadraticCurve
    slope: 1.0
    exponent: 0.4

# After
- !type:FoodValueCon
  hungerThreshold: Peckish # <--- Refers to a Satiation Value Key
  curve: !type:QuadraticCurve
    slope: 1.0
    exponent: 0.4
```

**Changelog**

No player-facing changes

:cl:
ADMIN:

- add: Satiation reduction command "unsatiate"
- remove: Hunger reduction command "hungry"
- remove: Thirst reduction command "thirst"
- add: Satiation setting command "setsatiation"
- remove: Hunger / thirst setting command "setnutrit"

Centronias committed:
ready

Centronias commented on Content.Client/Overlays/ShowSatiationIconsSystem.cs:
```diff
@@ -0,0 +1,41 @@
+
+    private void OnGetStatusIconsEvent(Entity<SatiationComponent> entity, ref GetStatusIconsEvent args)
+    {
+        // TODO I am not convinced that this `TryComp` is correct. It seems like it'll be looking for the component on
+        //  the entity whose status is being checked, rather than an entity who is viewing. Eg. you put on the beer
+        //  goggles or something and this system will be checking the PATRONS for the `can see thirst` component.
+        if (!IsActive || !TryComp<ShowSatiationIconsComponent>(entity, out var showComp))
+        {
+            return;
+        }
+
+        foreach (var shownTypeId in showComp.ShownTypes)
+        {
+            if (_satiation.GetStatusIconOrNull(entity, shownTypeId) is { } iconId)
+            {
+                args.StatusIcons.Add(iconId);
+            }
+        }
+    }
```
Oops I forgot about this. Gonna take a look.

slarticodefast commented:
Based PR. I'm so glad someone revived this.

Centronias committed:
oop forgot the HUD stuff

Alpaccalypse commented:
As someone living in .yml land, this looks great, even if I'll need to learn something new. I do know enough to say: wow, impressive work.

beck-thompson commented:
Eating and drinking combined and in shared? Would bring a tear to anyones eye

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Server/Nutrition/EntitySystems/FatExtractorSystem.cs
#	Content.Shared/Inventory/InventorySystem.Relay.cs

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Shared/Animals/UdderSystem.cs

Centronias committed:
woooo examinable generalized tooooo

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Shared/Nutrition/Components/ExaminableHungerComponent.cs
#	Content.Shared/Nutrition/EntitySystems/ExaminableHungerSystem.cs
#	Resources/Locale/en-US/nutrition/components/examineable-hunger-component.ftl
#	Resources/Prototypes/Entities/Mobs/NPCs/animals.yml

Centronias committed:
oop I forgot to rename the files

Centronias committed:
bungled the merge, I guess

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Server/Medical/VomitSystem.cs

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Server/NPC/Systems/NPCUtilitySystem.cs

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Resources/Prototypes/Entities/Mobs/Species/reptilian.yml
#	Resources/Prototypes/Reagents/Consumable/Drink/alcohol.yml

Centronias committed:
this is called hubris

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

VerinSenpai reviewed (changes requested):
I wish I could provide more insight but I'm a level 1 noob.

I hope a real reviewer can put your soul at ease :pray:

Centronias commented on Content.Server/Nutrition/Unsatiate.cs:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    // Always suggest any satiation types which aren't already in the arg list.
+    public override CompletionResult GetCompletion(IConsoleShell shell, string[] args) =>
```
It takes so much of my willpower to not make everything I can into an expression bodied function.

Centronias commented on Content.Server/Nutrition/Unsatiate.cs:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+
+    private void ExecuteWithoutArgs(IConsoleShell shell, Entity<SatiationComponent> entity)
```
I massaged it for a while and couldn't come up with anything definitely better.
Maybe if C# had better support for generic variance (ie. `entity.Comp.Satiations.Keys` and `args` could be implicitly inferred to one `IEnumerable<ProtoId>` sort of type) it could be better, but as it is now, there're manual conversions and new allocations needed for that which seem undesireable.

Centronias committed:
Verin's comments

iaada reviewed (commented):
I love the system and hope to see it finished. Here's a couple things I caught looking through.

Tayrtahn commented on Content.Shared/Nutrition/Prototypes/SatiationPrototype.cs:
```diff
@@ -0,0 +1,111 @@
+    #endregion
+
+
+    #region helpers
```
These are fine. Changing the state from the component would be bad, but pure getters are fine.

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Shared/Animals/UdderSystem.cs
#	Content.Shared/Animals/WoolySystem.cs
#	Content.Shared/Nutrition/EntitySystems/HungerSystem.cs
#	Content.Shared/Nutrition/EntitySystems/ThirstSystem.cs

Centronias commented on Content.Shared/EntityEffects/EffectConditions/Satiation.cs:
```diff
@@ -0,0 +1,53 @@
+            return false;
+        }
+
+        return satiation > Min && satiation < Max;
```
I agree, though [the previous implementation for `TotalHunger` had the same issue](https://github.com/Centronias/space-station-14/blob/a77c24cc864a926e3dc103ca8bf0a9996d6648e4/Content.Shared/EntityEffects/EffectConditions/TotalHunger.cs#L21).

iaada replied:
I have #36289 up to address the bug that came about from it, because yeah it's already caused a problem with Lipolicide.

Centronias commented on Content.Shared/Nutrition/Components/SatiationComponent.cs:
```diff
@@ -0,0 +1,52 @@
+    /// <summary>
+    /// This system handles pausing the <see cref="TimeSpan"/> fields in all of the values of <see cref="Satiations"/>.
+    /// </summary>
+    public sealed class PauseSystem : EntitySystem
```
I figured since nothing should ever actually reference the system, the name doesn't matter much, and even then, it would be qualified as `SatiationComponent.PauseSystem`, right? Admittedly I don't get C#'s nested classes too well, so maybe I'm wrong.

Either way, I think you're right that it should be part of the Satiation system. It's only here because I blindly looked at what `AutoGeneratePause` (or whatever it's called) makes, and it sticks the system in the component like this.

Centronias committed:
PR comments

how did I use `ProtoId<> Hunger` in more than a few spots but then just say "thirst can stay a literal :^)"?

also changed a few spots to my beloved
```
functionInvocation(
  one
  param
  per
  line
)
```
```

Centronias committed:
Satiation effect condition now with ADDED FUNCTIONALITY?? :o

Centronias committed:
tfw you trust the markov chain recommendation because long generic type names are tough q-q

Centronias committed:
TFW YOU TRUST THE MARKOV CHAIN BUT IT DOESN'T UNDERSTAND THE DIFFERING SEMANTICS OF MAX AND MIN AHHHHH

iaada commented on Content.Shared/Nutrition/EntitySystems/SatiationSystem.cs:
```diff
@@ -0,0 +1,330 @@
+        foreach (var satiation in entity.Comp.Satiations.Values)
+        {
+            var proto = _prototype.Index(satiation.Prototype);
+            var value = _random.Next(
```
I think you could get away with  just a comment explaining it, but a server check would fix the (really not very important) mispredict you'd get with the wrong alert for a couple seconds.

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

Centronias committed:
tfw you think to yourself "i need to fix the effects", but then forget to and the code doesn't even compile.

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Server/RatKing/RatKingSystem.cs

Princess-Cheeseballs commented on Content.Shared/Nutrition/EntitySystems/SatiationSystem.cs:
```diff
@@ -0,0 +1,368 @@
+        foreach (var satiation in entity.Comp.Satiations.Values)
+        {
+            var proto = _prototype.Index(satiation.Prototype);
+            var value = _random.Next(
```
This will mispredict exactly once at roundstart which isn't too big of an issue? But it still should be changed. Look at SharedBloodstreamSystem or SharedEntityEffectsSystem for an example on how to do predicted RNG.

Princess-Cheeseballs reviewed (changes requested):
Partial review I'm going to try and not forget to review it this time.

Centronias committed:
centralize satiation constants

Centronias commented on Content.Shared/Nutrition/EntitySystems/SatiationSystem.cs:
```diff
@@ -0,0 +1,368 @@
+        ref RefreshMovementSpeedModifiersEvent args
+    )
+    {
+        if (_jetpack.IsUserFlying(entity))
```
This was just uncritically ported [from `HungerSystem`](https://github.com/space-wizards/space-station-14/blob/master/Content.Shared/Nutrition/EntitySystems/HungerSystem.cs#L60)

<img width="445" height="235" alt="image" src="https://github.com/user-attachments/assets/8ea72e44-fb76-4995-9b29-6c16a877862e" />

If you're certain, I'm happy to remove it.

Princess-Cheeseballs replied:
100% good to remove. lmao.

Centronias committed:
other small comments

Centronias committed:
The enum is dead.
But at what cost?

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

Centronias commented on Content.IntegrationTests/Tests/Nutrition/HungerThirstTest.cs:
```diff
@@ -26,30 +25,34 @@ public async Task HungerThirstIncreaseDecreaseTest()
 
         // Ensure hunger and thirst value decrease over time (the Urist gets hungrier/thirstier)
-        var previousHungerValue = hungerSystem.GetHunger(hungerComponent);
-        var previousThirstValue = thirstComponent.CurrentThirst; // TODO: combined sation system with a sane API
```
> TODO: combined sation system with a sane API

unsure if sane api, please advise.

Centronias committed:
oop

Centronias committed:
like 12 more ooperses

Centronias commented:
Okay, description updated, tests tested, etc. etc.

Centronias committed:
green checks are mine by right; I have performed the rites; I have done my part; you will not deny me this succor

Centronias committed:
oops I changed the semantics from `<=` to just `<` and didn't update the tests :)

Princess-Cheeseballs commented on Resources/Prototypes/Body/Satiation/thirst.yml:
```diff
@@ -0,0 +1,75 @@
+  alerts:
+    Thirsty: Thirsty
+    Parched: Parched
+    Dead: Parched
```
Curious as to why parched is defined twice here but not twice for hunger?

Princess-Cheeseballs commented on Resources/Prototypes/Body/Satiation/hunger.yml:
Might be good to nest these in a folder for better organization so if more are added they don't make it harder to find types.yml

Princess-Cheeseballs commented on Content.Shared/Nutrition/Prototypes/SatiationPrototype.cs:
```diff
@@ -0,0 +1,194 @@
+
+    /// <inheritdoc/>
+    [AbstractDataField]
+    public bool Abstract { get; }
```
If you're going to allow for abstract types, you should have hunger and thirst parent off of an abstract satiation prototype.

Centronias replied:
tbh I just copied this from whatever I based the "how do I make a prototype" code. I looked at them and there's not really anything shareable on a parent between hunger and thirst, so I'll just get rid of allowing them to be abstract.

Princess-Cheeseballs reviewed (changes requested):
Very impressive PR. Sorry it's taken a bit to review it. I need to do a bit more testing with it but here are a few things I noticed.

Centronias committed:
serialization has bested me :agony:

also PR comments

Centronias committed:
Merge remote-tracking branch 'upstream/master' into satiation

# Conflicts:
#	Resources/Prototypes/Entities/Mobs/NPCs/regalrat.yml

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Resources/Prototypes/Reagents/narcotics.yml

Centronias committed:
never trust the auto merge :agony:

Centronias committed:
Merge branch 'master' into satiation

# Conflicts:
#	Resources/Prototypes/Entities/Clothing/Eyes/hud.yml
#	Resources/Prototypes/Entities/Mobs/NPCs/regalrat.yml
#	Resources/Prototypes/Entities/Mobs/Species/arachnid.yml
#	Resources/Prototypes/Entities/Mobs/Species/diona.yml
#	Resources/Prototypes/Entities/Mobs/Species/dwarf.yml
#	Resources/Prototypes/Entities/Mobs/Species/human.yml
#	Resources/Prototypes/Entities/Mobs/Species/moth.yml
#	Resources/Prototypes/Entities/Mobs/Species/reptilian.yml
#	Resources/Prototypes/Entities/Mobs/Species/slime.yml
#	Resources/Prototypes/Entities/Mobs/Species/vox.yml
#	Resources/Prototypes/Entities/Mobs/Species/vulpkanin.yml
#	Resources/Prototypes/Reagents/Consumable/Drink/alcohol.yml
#	Resources/Prototypes/Reagents/Consumable/Drink/drinks.yml
#	Resources/Prototypes/Reagents/Consumable/Drink/juice.yml
#	Resources/Prototypes/Reagents/Consumable/Food/ingredients.yml

Centronias committed:
literally just this

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Server/NPC/Systems/NPCUtilitySystem.cs

Centronias committed:
force engine back to master because idk why it's different

Princess-Cheeseballs committed:
Merge branch 'master' into fork/Centronias/satiation

Princess-Cheeseballs committed:
fix randomness and update master

Princess-Cheeseballs reviewed (approved):
Sorry for the exceptionally long review time. Finally found some time to do a proper review, I merged master and updated the random extension you used in satiation system. From testing and code review looks good, I thank you for your efforts.

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Server/NPC/Systems/NPCUtilitySystem.cs

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Server/NPC/Systems/NPCUtilitySystem.cs
#	Content.Shared/Nutrition/EntitySystems/ExaminableHungerSystem.cs

themias commented:
Related to (and prerequisite for) #18824

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Client/Overlays/ShowHungerIconsSystem.cs
#	Content.Client/Overlays/ShowThirstIconsSystem.cs
#	Content.Server/Animals/Systems/EggLayerSystem.cs
#	Content.Server/NPC/HTN/Preconditions/HungryPrecondition.cs
#	Content.Server/NPC/HTN/Preconditions/ThirstyPrecondition.cs
#	Content.Server/NPC/Systems/NPCUtilitySystem.cs
#	Content.Server/Nutrition/EntitySystems/AnimalHusbandrySystem.cs
#	Content.Server/Nutrition/EntitySystems/FatExtractorSystem.cs
#	Content.Server/RatKing/RatKingSystem.cs
#	Content.Shared/Animals/UdderSystem.cs
#	Content.Shared/Animals/WoolySystem.cs
#	Content.Shared/EntityConditions/Conditions/Body/HungerEntityConditionSystem.cs
#	Content.Shared/EntityEffects/Effects/Body/SatiateEntityEffectSystem.cs
#	Content.Shared/Medical/VomitSystem.cs
#	Content.Shared/Nutrition/EntitySystems/ExaminableHungerSystem.cs
#	Content.Shared/Nutrition/EntitySystems/HungerSystem.cs
#	Content.Shared/Nutrition/EntitySystems/ThirstSystem.cs
#	Content.Shared/Sericulture/SericultureSystem.cs
#	Resources/Prototypes/Entities/Mobs/NPCs/animals.yml
#	Resources/Prototypes/Entities/Mobs/NPCs/space.yml

Centronias committed:
idk why whitelist system is complaining now, but here you go

Centronias committed:
submodule fuckery

Centronias committed:
fall before my might

Centronias committed:
okay I ran the linter, ARE YOU HAPPY??

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

Centronias committed:
cursed ovis, trouble me no more

AJCM-git reviewed (changes requested):
Same reviews as https://github.com/Space-Wizards-Federation/space-station-14/pull/61

If you have questions/want to discuss a review quote and post it here as a comment.

whatston3 reviewed (commented):
Pardon the merge conflicts.

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Shared/Medical/VomitSystem.cs
#	Content.Shared/Nutrition/EntitySystems/HungerSystem.cs
#	Content.Shared/Nutrition/EntitySystems/ThirstSystem.cs

Centronias committed:
protoman stuff

Centronias reviewed (commented):
@AJCM-git , I went through and dealt with or responded to your comments from https://github.com/Space-Wizards-Federation/space-station-14/pull/61 . Lemme know if I missed anything.

> consider a rebranding

I don't have any other good name beyond like `PhysicalNeeds`, which doesn't seem much better to me. I guess it's more clear that it's more flexible.

> over complicating some things for (i assume) negligible perf gains

I haven't profiled anything, I just went off of gut based on an original implementation that was doing a ton of collating and lookups through the raw prototype info. My intention was that the only complicated-for-perf stuff would be the caching which happens once (or when prototypes are reloaded), is contained entirely within one file, and has its benefits exposed by a clear API. So I will concede that it may be premature optimization, but I don't think it's a huge cost beyond the cost of maintaining the implementation.

Maybe it'd be worth it to define a data structure type that implements the Threshold behavior with a clearer API to better deal with it? Probably would get a builder, too, so that it's really easy to use. Something to consider.

Centronias committed:
inb4 linter bonks me

AJCM-git reviewed (changes requested):
After sitting on it for a bit the PhysicalNeeds name did grow on me, feel free to rebrand to that. Again not a requirement for this to be accepted, just something you can do to make the goals of this system slightly clearer

Centronias committed:
break non-critical stuff from the main satiation proto

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Shared/EntityConditions/Conditions/Body/HungerEntityConditionSystem.cs
#	Resources/Prototypes/Entities/Mobs/NPCs/regalrat.yml

Centronias committed:
![live crundle reaction](https://preview.redd.it/xh68xqcz518g1.png?width=677&format=png&auto=webp&s=79ba10fe49ec3f8306217732321a7c5429bdfb14)

inb4 this breaks something somewhere lmao

Centronias committed:
my two greatest PRs COLLIDE

I wonder why robust says it's changed :hmmclueless

Centronias committed:
sanity is at an all-time low

Centronias committed:
serialization, I guess

Centronias committed:
I don't believe you

Centronias committed:
Merge remote-tracking branch 'upstream/master' into satiation

AJCM-git commented on Content.Server/Nutrition/EntitySystems/AnimalHusbandrySystem.cs:
```diff
@@ -148,10 +158,26 @@ public bool CanReproduce(EntityUid uid, ReproductiveComponent? component = null)
+            (
+                !satiation.Has(SatiationSystem.Thirst) ||
+                _satiation.IsValueInRange((uid, satiation), SatiationSystem.Thirst, below: minThirst)
+            ))
```
weird formatting. possible helper method for this?

AJCM-git reviewed (changes requested):
Cursory glance but this is pretty close to done
DirtyFields now no longer kills you if your component only has 1 field, i would say switching most Dirty to it would be warranted since the satiation components look like the kind of thing other people will add more fields to in the future so better to future-proof

SlamBamActionman reviewed (dismissed):
Halfway review

Centronias committed:
Merge remote-tracking branch 'origin/master' into satiation

# Conflicts:
#	Content.Shared/Nutrition/EntitySystems/ThirstSystem.cs
#	Resources/Prototypes/Body/species_base.yml
#	Resources/Prototypes/Entities/Mobs/NPCs/animals.yml
#	Resources/Prototypes/Entities/Mobs/NPCs/regalrat.yml
#	Resources/Prototypes/Reagents/Consumable/Drink/alcohol.yml

Centronias committed:
field deltas on class \w*satiation\w*components

Centronias committed:
PR comments

Centronias committed:
我做泥團子 :pensive:

AJCM-git reviewed (approved):
api could be refined a bit more but we can iterate in the future because this is a very solid base.
someone remind me to make an issue about rebranding this to something more general after merging

SlamBamActionman commented on Content.Shared/Nutrition/Prototypes/SatiationPrototype.cs:
```diff
@@ -0,0 +1,148 @@
+    [DataField("startingValueMaximum", required: true)]
+    private SatiationValue _startingValueMaximum = default!;
+
+    public int StartingValueMaximum => GetValueOrNull(_startingValueMaximum) ?? MaximumValue;
```
Ditto

SlamBamActionman reviewed (changes requested):
Second half review (really just three places for XML comments). 

I am going on the record to say that I am slightly sus on the BaseSatiationEffectSystem and its implementation. It's done well, but I am feeling we may be risking some overlap with EntityEffect/StatusEffect systems. However I understand the need for it in certain cases (`ExaminableSatiationSystem` is one such good example), and it's possible that the satiation effect system could be hooked into the aforementioned effect systems at a later date.

Otherwise I can't glean any issues from it - great job.

Centronias committed:
doc + kill one property that's trivially accessible after the threshold caching changes

SlamBamActionman reviewed (approved):
Crossing our fingers everything works out. To Vulture, and beyond!

AJCM-git committed:
Merge branch 'master' into pr/34166

Jezithyr reviewed (approved):
This looks *sick*, nice work!

AJCM-git committed:
Merge branch 'master' into pr/34166

AJCM-git committed:
fixes
